### PR TITLE
Add command to force tags on nodes and list tags

### LINKED ISSUE #356: View Tags
Is there any way to view tags of hosts? I'm on the current version.

### LINKED ISSUE #525: Manually add tags to nodes from headscale
It's been reported on discord that users would need the ability to add tags to nodes from headscale.

### PR DESCRIPTION
**Pull request by restanrm**
- [x] read the [CONTRIBUTING guidelines](README.md#user-content-contributing)
- [x] raised a GitHub issue or discussed it on the projects chat beforehand
- [x] added unit tests
- [x] added integration tests
- [ ] updated documentation if needed
- [x] updated CHANGELOG.md

This PR will fix #356 and fix #525

Work to be done before finalisation of this PR
- [x] Handle validation errors or normalisation of the name of the tags (tags must start with the `tag:` prefix)
  this means that we should define a way to return errors in the protobuf. Should we define status and message ?
- [x] replace add/del in CLI by set
- [x] make a generic `contains` function
- [x] change `updateDBMachine` to `UpdateTags`
- [x] add integration tests
- [ ] add documentation

image

### CODE REVIEW

File: machine.go [hunk 1]
```diff
@@ -641,6 +661,41 @@ func (machine *Machine) toProto() *v1.Machine {
...
+// getTags will return the tags of the current machine.
+func getTags(
+	aclPolicy ACLPolicy,
+	machine Machine,
+	stripEmailDomain bool,
+) (validTags []string, invalidTags []string) {
+	validTagMap := make(map[string]bool)
+	invalidTagMap := make(map[string]bool)
+	for _, tag := range machine.HostInfo.RequestTags {
+		owners, err := expandTagOwners(aclPolicy, tag, stripEmailDomain)
```
Comment (kradalby) [inline machine.go:673]: This whole part of the invalidTagMap really confuses me, and I cant really test what happens from the tests, could you try to elaborate?

**Review by kradalby — COMMENTED** [review]
I think this looks quite reasonable, some questions, some design.

**Comment by kradalby** [comment]
And we will want some integration tests

File: machine.go [hunk 2]
```diff
@@ -357,6 +359,22 @@ func (h *Headscale) UpdateMachine(machine *Machine) error {
 	return nil
 }
 
+// UpdateDBMachine takes a Machine struct pointer (typically already loaded from database
+// search for the same machine in the database and update the latter
+func (h *Headscale) UpdateDBMachine(machine Machine) error {
```
[the remark this answers is not in the document]
Comment (restanrm) [reply machine.go:365]: We need a different name than `UpdateMachine` that is already used and update from DB to local. In this function we want opposite behavior.

[the remark this answers is not in the document]
Comment (restanrm) [reply machine.go:365]: I'm open to better names of course !

File: acls_test.go [hunk 3]
```diff
@@ -1017,6 +1017,44 @@ func Test_expandAlias(t *testing.T) {
...
+			name: "Forced tag defined",
+			args: args{
+				alias: "tag:hr-webserver",
+				machines: []Machine{
+					{
+						IPAddresses: MachineAddresses{
+							netaddr.MustParseIP("100.64.0.1"),
+						},
+						Namespace:  Namespace{Name: "joe"},
+						ForcedTags: []string{"tag:hr-webserver"},
```
[the remark this answers is not in the document]
Comment (restanrm) [reply acls_test.go:1030]: Currently, the user need to pass `tag:hr-webserver` as for the login in `tailscale up --advertise-tags` I need to add validation in server implementation and add proper messages in protobuf

File: machine.go (hunk 1 above)
Comment (restanrm) [reply machine.go:673]: What is bothering you ? The fact that I use a map or the concept of invalidTags ? I'm using a map to remove duplicated tags at insertion. It is possible for a machine to have multiple times the same tag (declared in `tailscale up`).

File: machine.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (kradalby) [reply machine.go:365]: Ah I see, How about having a method that does the one thing instead? `SetTags` or something, and not have it arbitrarily update some of the fields on the machine object?

File: gen/go/headscale/v1/headscale.pb.go [hunk 4]
```diff
@@ -36,7 +36,7 @@ var file_headscale_v1_headscale_proto_rawDesc = []byte{
 	0x6f, 0x74, 0x6f, 0x1a, 0x19, 0x68, 0x65, 0x61, 0x64, 0x73, 0x63, 0x61, 0x6c, 0x65, 0x2f, 0x76,
```
[the remark this answers is not in the document]
Comment (restanrm) [reply gen/go/headscale/v1/headscale.pb.go:36]: Ok will do !

File: cmd/headscale/cli/tags.go [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+}
+
+var tagCmd = &cobra.Command{
+	Use:     "tags",
+	Short:   "Manage the tags of Headscale",
+	Aliases: []string{"t", "tag"},
+}
+
+var addTagCmd = &cobra.Command{
+	Use:   "add",
```
[the remark this answers is not in the document]
Comment (restanrm) [reply cmd/headscale/cli/tags.go]: That's a good proposition, I'll update to do this !

File: machine.go (hunk 1 above)
Comment (restanrm) [reply machine.go:673]: Right, so an invalid tag is a tag added by a user that does'nt have permission to add tags. If joe (a simple user) can setup his computer to be a server, then he can talk to the database directly even if he should not. So the validation of the tags are made with the TagOwners in the ACL file. The documentation of this feature has not been made yet. It seems necessary.

File: grpcv1.go [hunk 6]
```diff
@@ -182,6 +184,37 @@ func (api headscaleV1APIServer) GetMachine(
...
+) (*v1.UpdateMachineResponse, error) {
+	rMachine := request.GetMachine()
+	machine, err := api.h.GetMachineByID(rMachine.Id)
+	if err != nil {
+		return nil, err
+	}
+
+	machine.ForcedTags = rMachine.ForcedTags
+	machine.Name = rMachine.Name
+	id, err := strconv.Atoi(rMachine.Namespace.Id)
```
[the remark this answers is not in the document]
Comment (restanrm) [reply grpcv1.go]: code removed

File: machine.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (restanrm) [reply machine.go]: done

File: cmd/headscale/cli/tags.go [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+				updateResponse.GetMachine(),
+				"Machine updated",
+				output,
+			)
+		}
+
+	},
+}
+
+func containsString(ss []string, s string) bool {
```
[the remark this answers is not in the document]
Comment (restanrm) [reply cmd/headscale/cli/tags.go]: changed

File: cmd/headscale/cli/tags.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (restanrm) [reply cmd/headscale/cli/tags.go]: done

File: cmd/headscale/cli/tags.go [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+			)
+		}
+
+	},
+}
+
+var delTagCmd = &cobra.Command{
+	Use:     "del",
+	Short:   "remove tags to a node in your network",
+	Aliases: []string{"remove", "rm"},
```
[the remark this answers is not in the document]
Comment (restanrm) [reply cmd/headscale/cli/tags.go]: command removed

File: machine.go (hunk 1 above)
Comment (restanrm) [reply machine.go]: I feel that we need to explain this somewhere. Do you have any good location for this ?

File: utils.go [hunk 9]
```diff
@@ -282,9 +273,9 @@ func stringToIPPrefix(prefixes []string) ([]netaddr.IPPrefix, error) {
 	return result, nil
 }
 
-func containsIPPrefix(prefixes []netaddr.IPPrefix, prefix netaddr.IPPrefix) bool {
-	for _, p := range prefixes {
-		if prefix == p {
+func contains[T string | netaddr.IPPrefix](ts []T, t T) bool {
```
Comment (kradalby) [inline utils.go:276]: nice!

File: machine.go (hunk 1 above)
Comment (kradalby) [reply machine.go]: hmm, maybe in the signature of the function 🤔

**Review by kradalby — COMMENTED** [review]
I think this looks reasonable. Happy to get this in when things passes :)

File: utils.go (hunk 9 above)
Comment (restanrm) [reply utils.go:276]: Happy to change this :D

**Comment by kradalby** [comment]
Let's merge this, but can you open a pr or issue with the golang ci snippets that's removed to we can debug that separately ?

I don't want to remove it, but we can remove and then add it back in.

**Review by kradalby — APPROVED** [review]

**Comment by restanrm** [comment]
> Let's merge this, but can you open a pr or issue with the golang ci snippets that's removed to we can debug that separately ?
> 
> I don't want to remove it, but we can remove and then add it back in.

You mean the version pinning in the GithubAction and the ignored linters ?

**Comment by kradalby** [comment]
> You mean the version pinning in the GithubAction and the ignored linters ?

https://github.com/juanfont/headscale/commit/9f082125fa8e6674d2b08e47dbf48949cedef53c

This one

**Comment by restanrm** [comment]
oh, ok it didn't work anyway…

**Comment by blargg** [comment]
`headscale node list --tags` will list the nodes with tag information. It's not shown in the default node list (without `--tags`).

**Pull request merged**